### PR TITLE
Account store account type crash fix

### PR DESCRIPTION
**Pull request by bmctigue**
This pull request fixes issue #182.

### CODE REVIEW

File: Sources/OAuth2Client/NXOAuth2AccountStore.m [hunk 1]
```diff
@@ -545,12 +547,8 @@ - (void)oauthClientDidRefreshAccessToken:(NXOAuth2Client *)client
     }
     
     foundAccount.accessToken = client.accessToken;
-    NSDictionary *userInfo = [NSDictionary dictionaryWithObject: foundAccount
-                                                         forKey: NXOAuth2AccountStoreNewAccountUserInfoKey];
-    
-    [[NSNotificationCenter defaultCenter] postNotificationName:NXOAuth2AccountStoreAccountsDidChangeNotification
-                                                        object:self
-                                                      userInfo:userInfo];
+    [[NSNotificationCenter defaultCenter] postNotificationName:NXOAuth2AccountDidChangeAccessTokenNotification
```
Comment (toto) [inline Sources/OAuth2Client/NXOAuth2AccountStore.m:550]: Why did you change the notification. This is part of the account store and `NXOAuth2AccountStoreAccountsDidChangeNotification` is the correct notification type. In any case this is not part of the crash fix.

Comment (bmctigue) [inline Sources/OAuth2Client/NXOAuth2AccountStore.m]: I added the same correction here, because it has the same vulnerability. Why not prevent the same issue in the notification.

Comment (bmctigue) [inline Sources/OAuth2Client/NXOAuth2AccountStore.m]: But your right, if this can never happen in the notification, perhaps the change is unnecessary.

File: Sources/OAuth2Client/NXOAuth2Connection.m [hunk 2]
```diff
@@ -418,7 +418,7 @@ - (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLRespon
     if (self.statusCode == 401
         && client.accessToken.refreshToken != nil
         && authenticateHeader
-        && [authenticateHeader rangeOfString:@"expired_token"].location != NSNotFound) {
+        && ([authenticateHeader rangeOfString:@"invalid_token"].location != NSNotFound || [authenticateHeader rangeOfString:@"invalid_request"].location != NSNotFound)) {
```
Comment (toto) [inline Sources/OAuth2Client/NXOAuth2Connection.m:421]: This does not look right: The case is a 401 with expiration explicitly because we want to refresh the token in that case. The other cases need to be covered somewhere else

Also this is not part of the crash fix as well.

File: Sources/OAuth2Client/NXOAuth2Connection.m [hunk 3]
```diff
@@ -469,7 +469,7 @@ - (void)connectionDidFinishLoading:(NSURLConnection *)connection;
                 }
             }
             if (authenticateHeader
-                && [authenticateHeader rangeOfString:@"invalid_token"].location != NSNotFound) {
+                && ([authenticateHeader rangeOfString:@"invalid_token"].location != NSNotFound || [authenticateHeader rangeOfString:@"invalid_request"].location != NSNotFound)) {
```
Comment (toto) [inline Sources/OAuth2Client/NXOAuth2Connection.m:472]: :+1:

File: Sources/OAuth2Client/NXOAuth2AccountStore.m (hunk 1 above)
Comment (bmctigue) [inline Sources/OAuth2Client/NXOAuth2AccountStore.m:550]: Sorry, I was attempting to make the calls consistent. The access token has changed, so I used this notification type. I understand the account has changed also.

File: Sources/OAuth2Client/NXOAuth2Connection.m (hunk 2 above)
Comment (bmctigue) [inline Sources/OAuth2Client/NXOAuth2Connection.m:421]: Maybe, but it has permanently fixed our crashes. We have had not issues since implementing this. I understand it may not be correct generally, but we had to get this working. Thanks for looking at it!!

Comment (toto) [inline Sources/OAuth2Client/NXOAuth2Connection.m:421]: This feels like a server problem maybe. What this does it checks if the request was rejected as unauthenticated and refreshes the token if that is the case. If you change it the way you did auto-token refreshing is broken (which you might not notice if your access tokens do not expire). 

If this change fixed it maybe your server sends 401 in an unexpected way (note that 401 responses without `expired_token` should be handled by your change below, so maybe this line need not be changed?

Comment (bmctigue) [inline Sources/OAuth2Client/NXOAuth2Connection.m:421]: Thanks! I'm the iPhone side of this, so I can't speak to the server side code. I'll pass this along to our dev. We were testing against a server that expired the access token every 60 seconds.

Comment (toto) [inline Sources/OAuth2Client/NXOAuth2Connection.m:421]: Mmm, could you double check the HTTP Response in case of the expiration.

Comment (bmctigue) [inline Sources/OAuth2Client/NXOAuth2Connection.m:421]: Here's the authentication header in response to a 401, the debugger is paused at line 422:
Printing description of authenticateHeader:
Bearer realm="BetterUp", error="invalid_token", error_description="The access token is invalid"

Comment (toto) [inline Sources/OAuth2Client/NXOAuth2Connection.m:421]: `invalid_token` is not listed in the [error codes of the RFC6749](http://tools.ietf.org/html/rfc6749#section-5.2) Curiously `expired_token` also seems wrong (it was correct for [Draft 10](http://tools.ietf.org/html/draft-ietf-oauth-v2-10), when this was built). `invalid_client` seems to be the correct error response together with a 401 (see RFC linked above). Can you find out which client library they are using?

**Comment by sirnacnud** [comment]
Would it be possible just to merge the crash fix in?

**Comment by toto** [comment]
I merged it but removed `invalid_client` from automatic token refresh. See the commit message from  0385fa1

**Comment by sirnacnud** [comment]
I'm not for sure what exactly happened, but cad25e4 has a failed merged in NXOAuth2Connection.m starting at line 478.
